pkg/destroy/aws/aws.go

```diff
@@ -77,1510 +63,1205 @@ func (o *ClusterUninstaller) validate() error {
+	segments := strings.SplitN(roleARN.Resource, "/", 2)
+	if len(segments) != 2 {
+		return errors.Errorf("cannot parse resource %q into {type}/{id}", roleARN.Resource)
+	}
+	resourceType := segments[0]
+	name := segments[1]
+	logger = logger.WithField("name", name)
+
+	if resourceType != "role" {
+		return errors.Errorf("%s ARN passed to deleteIAMRole: %s", resourceType, roleARN.String())
+	}
 
-	var result *route53.ListTagsForResourceOutput
 	var err error
-	for _, zone := range zones {
-		for {
-			result, err = r53Client.ListTagsForResource(&route53.ListTagsForResourceInput{
-				ResourceType: aws.String("hostedzone"),
-				ResourceId:   zone.Id,
-			})
-			if err != nil {
-				if request.IsErrorThrottle(err) {
-					logger.Debugf("sleeping before trying to resolve tags for zone %s: %v", *zone.Id, err)
-					time.Sleep(time.Second)
-					continue
+	err2 := client.ListRolePoliciesPages(
+		&iam.ListRolePoliciesInput{RoleName: &name},
+		func(results *iam.ListRolePoliciesOutput, lastPage bool) bool {
+			for _, policy := range results.PolicyNames {
+				_, err = client.DeleteRolePolicy(&iam.DeleteRolePolicyInput{
+					RoleName:   &name,
+					PolicyName: policy,
+				})
+				if err != nil {
+					err = errors.Wrapf(err, "deleting IAM role policy %s", *policy)
+					logger.Info(err)
 				}
-				return zonesAsAWSObjects, err
+				logger.WithField("policy", *policy).Info("Deleted")
 			}
 
-			break
-		}
-
-		tagsToMap, err := tagsToMap(result.ResourceTagSet.Tags)
-		if err != nil {
-			return zonesAsAWSObjects, err
-		}
-
-		zonesAsAWSObjects = append(zonesAsAWSObjects, awsObjectWithTags{
-			Name: *zone.Id,
-			Tags: tagsToMap,
-		})
+			return !lastPage
+		},
+	)
 
+	if err != nil {
+		return err
+	}
+	if err2 != nil {
+		return errors.Wrap(err2, "listing IAM role policies")
 	}
 
-	return zonesAsAWSObjects, nil
-}
+	err2 = client.ListInstanceProfilesForRolePages(
+		&iam.ListInstanceProfilesForRoleInput{RoleName: &name},
+		func(results *iam.ListInstanceProfilesForRoleOutput, lastPage bool) bool {
+			for _, profile := range results.InstanceProfiles {
+				parsed, err := arn.Parse(*profile.Arn)
+				if err != nil {
+					err = errors.Wrap(err, "parse ARN for IAM instance profile")
+					logger.Info(err)
+					continue
+				}
 
-// deleteEntriesFromSharedR53Zone will find route53 entries for the shared (ie non-terraform-managed) route53 zone
-// and remove them.
-// Provide the terraform-created private zone, and the manually created public/shared zone, and it will find any
-// entries in the public/shared zone that match entries in the private zone, and delete them
-func deleteEntriesFromSharedR53Zone(zoneID string, sharedZoneID string, r53Client *route53.Route53, logger logrus.FieldLogger) error {
+				err = deleteIAMInstanceProfile(client, parsed, logger.WithField("IAM instance profile", parsed.String()))
+				if err != nil {
+					err = errors.Wrapf(err, "deleting %s", parsed.String())
+					logger.Info(err)
+				}
+			}
+
+			r
```

Comment:
this and https://github.com/openshift/installer/pull/1039/commits/5873630befa7f5921d4bee58290070aea1084233#diff-ab65f0e2ba0237c5cc1429673c22cc08R1188 should be using the same format / function ?

Reply:
>  I use that twice within 20 lines in the same function, but I can pull it out into a helper function if you're worried about drift between those two call sites.
I know it is close, but just pulling it out into inline func would be enough too..
```go
func deleteRoute53(session *session.Session, arn arn.ARN, logger logrus.FieldLogger) error {
...
keyFunc := func (rset *ResourceRecordSets) {
    key := fmt.Sprintf("%s %s", *rset.Type, *rset.Name)
}
...
}
```

Reply:
> I know it is close, but just pulling it out into inline func...

Done with 122e743 -> da9b6c9.